pkg/apis/resource/types.go

```diff
@@ -0,0 +1,466 @@
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/apis/core"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Claim is created by users to describe which resources they need.
+// Its status tracks whether the resource has been allocated and what the
+// resulting attributes are.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type Claim struct {
+	metav1.TypeMeta
+
+	// The driver must set a finalizer here before it attempts to allocate
+	// the resource. It removes the finalizer again when a) the allocation
+	// attempt has definitely failed or b) when the allocated resource was
+	// freed. This ensures that resources are not leaked.
+	metav1.ObjectMeta
+
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// Claim.
+	Spec ClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ClaimStatus
+}
+
+// ClaimSpec defines how a resource is to be allocated.
+type ClaimSpec struct {
+	// ClassName references the driver and additional parameters
+	// via the name of a Class that was created as part of the
+	// driver deployment.
+	//
+	// The apiserver does not check that the referenced class exists.
+	ClassName string
+
+	// ParametersRef references a separate object with arbitrary parameters
+	// that will be used by the driver when allocating a resource for the
+	// claim.
+	//
+	// The object must be in the same namespace as the Claim.
+	ParametersRef *ClaimParametersReference
+
+	// Allocation can start immediately or when a Pod wants to use the
+	// resource. "WaitForFirstConsumer" is the default.
+	AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a Claim gets allocated immediately
+// when it gets created (AllocationModeImmediate) or whether allocation is
+// delayed until it is needed for a Pod
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a Claim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the Claim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a Claim has AllocationModeImmediate, allocation starts
+	// as soon as the Claim gets created. This is done without
+	// considering the needs of Pods that will use the Claim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ClaimStatus struct {
+	// DriverName is a copy of the driver name from the Class at
```

Comment:
I think the API docs should describe what it is, but not "it's necesssary for..." sorts of things

Reply:
Reduced to just the first sentence.